drivers/note/notesched_driver.c

```diff
@@ -0,0 +1,393 @@
+{
+  "NOTE_START",
+  "NOTE_STOP",
+  "NOTE_SUSPEND",
+  "NOTE_RESUME",
+  "NOTE_CPU_START",
+  "NOTE_CPU_STARTED",
+  "NOTE_CPU_PAUSE",
+  "NOTE_CPU_PAUSED",
+  "NOTE_CPU_RESUME",
+  "NOTE_CPU_RESUMED",
+  "NOTE_PREEMPT_LOCK",
+  "NOTE_PREEMPT_UNLOCK",
+  "NOTE_CSECTION_ENTER",
+  "NOTE_CSECTION_LEAVE",
+  "NOTE_SPINLOCK_LOCK",
+  "NOTE_SPINLOCK_LOCKED",
+  "NOTE_SPINLOCK_UNLOCK",
+  "NOTE_SPINLOCK_ABORT",
+  "NOTE_SYSCALL_ENTER",
+  "NOTE_SYSCALL_LEAVE",
+  "NOTE_IRQ_ENTER",
+  "NOTE_IRQ_LEAVE",
+  "NOTE_DUMP_STRING",
+  "NOTE_DUMP_BINARY"
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: notesched_common
+ ****************************************************************************/
+
+static inline void notesched_common(FAR struct tcb_s *tcb, uint8_t type,
+                                    uintptr_t args)
+{
+  /* atomic operation, equivalent to g_notesched.index++ */
+
+  size_t index = atomic_fetch_add(&g_notesched.index, 1);
+  FAR struct notesched_chunk_s *note = &g_notesched.buffer[index];
+  uint32_t freq = up_perf_getfreq();
+
+  if (freq != 0)
+    {
+      uint32_t tick = up_perf_gettime();
+      note->time.tv_nsec = (uint64_t)tick * NSEC_PER_SEC / freq;
+      note->time.tv_sec = tick / freq;
+    }
+  else
+    {
+      clock_systime_timespec(&note->time);
+    }
+
+  note->args = args;
+  note->type = type;
+  if (tcb)
+    {
+      note->pid = tcb->pid;
+#ifdef CONFIG_SMP
+      note->cpu = tcb->cpu;
+#endif
+    }
+
+  if (g_notesched.index == CONFIG_DRIVER_NOTESCHED_NBUFFERS)
+    {
+      /* atomic operation, equivalent to g_notesched.index = 0 */
+
+      atomic_exchange(&g_notesched.index, 0);
+    }
+}
+
+/****************************************************************************
+ * Name: notesched_*
+ ****************************************************************************/
+
+static void notesched_start(FAR struct note_driver_s *drv,
+                            FAR struct tcb_s *tcb)
+{
+  notesched_common(tcb, NOTE_START, 0);
+}
+
+static void notesched_stop(FAR struct note_driver_s *drv,
+                           FAR struct tcb_s *tcb)
+{
+  notesched_common(tcb, NOTE_STOP, 0);
+}
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SWITCH
+static void notesched_suspend(FAR struct note_driver_s *drv,
+                              FAR struct tcb_s *tcb)
+{
+  notesched_common(tcb, NOTE_SUSPEND, 0);
+}
+
+static void notesched_resume(FAR struct note_driver_s *drv,
+                             FAR struct tcb_s *tcb)
+{
+  notesched_common(tcb, NOTE_RESUME, 0);
+}
+#endif
+
+#ifdef CONFIG_SMP
+static void notesched_cpu_start(FAR struct note_driver_s *drv,
+                                FAR struct tcb_s *tcb, int cpu)
+{
+  notesched_common(tcb, NOTE_RESUME, (uintptr_t)cpu);
```

Comment:
why use NOTE_RESUME, not NOTE_CPU_START?

Reply:
done